index.d.ts

```diff
@@ -0,0 +1,125 @@
+      => State<S>
+      | [State<S>, ...EffectDescriptor<D>[]]
+      | Action<S, P, D>)
+
+  // A payload is data external to state that is given to a dispatched action.
+  type Payload<P> = P
+
+  // An effect descriptor describes how an effect should be invoked.
+  // A function that creates this is called an effect constructor.
+  type EffectDescriptor<D> = [Effect, EffectData<D>]
+
+  // An effect is where side effects and any additional dispatching occur.
+  type Effect = <D>(dispatch: Dispatch, props?: EffectData<D>) => void
+
+  // An effect is generally given additional data.
+  type EffectData<D> = D
+
+  // ---------------------------------------------------------------------------
+
+  // A virtual DOM node represents an actual DOM element.
+  type VDOM = {
+    readonly type: string;
+    readonly props: PropList;
+    readonly children: VNode[];
+    node: MaybeNode;
+    readonly tag?: Tag;
+    readonly key: Key;
+    memo?: PropList;
+  }
+
+  // Virtual DOM properties will often correspond to HTML attributes.
+  type Prop = bigint | boolean | null | number | string | symbol | undefined | Function | ClassProp | StyleProp
+  type PropList = Readonly<ElementCreationOptions & {
+    [k: string]: Prop;
+    class?: ClassProp;
+    key?: Key;
+    style?: StyleProp;
+  }>
+
+  // A key can uniquely associate a virtual DOM node with a certain DOM element.
+  type Key = null | string | undefined
+
+  // The `class` property represents an HTML class attribute string.
+  type ClassProp = string | Record<string, boolean> | ClassProp[]
+
+  // The `style` property represents inline CSS.
+  type StyleProp = Record<string, null | number | string>
+
+  // A virtual node is a convenience layer over a virtual DOM node.
+  type VNode = null | undefined | VDOM
+
+  // Actual DOM nodes will be manipulated depending on how property patching goes.
+  type MaybeNode = null | undefined | Node
+
+  // A virtual DOM node's tag has metadata relevant to it. Virtual DOM nodes are
+  // tagged by their type to assist rendering.
+  type Tag = VDOMNodeType | View
+
+  // These are based on actual DOM node types:
+  // https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType
+  const enum VDOMNodeType {
+    SSR = 1,
+    Text = 3,
+  }
+
+  // ---------------------------------------------------------------------------
+
+  // The `app` function initiates a Hyperapp application. `app` along with effects
+  // should be the only places you need to worry about side effects.
+  function app<S, P, D>(props: App<S, P, D>): Dispatch
+
+  // The `h` function builds a virtual DOM node.
+  function h(type: string, props: PropList, children?: VNode | readonly VNode[]): VDOM
+
+  // The `memo` function stores a view along with properties for it.
+  function memo(view: View, props: PropList): Partial<VDOM>
+
+  // The `text` function creates a virtual DOM node representing plain text.
+  function text(value: string | number, node?: Node): VDOM
```

Comment:
@icylace +1 to the type definition! 

Small suggestion, the `value` argument on the text() function also accepts an array of strings, numbers, booleans, null (but not undefined) and it will convert this array into a comma-separated string.

Reply:
I explored `text` a bit more and found that only symbols give it trouble:
https://codepen.io/icylace/pen/XWXoLYy?editors=0010

Arrays of undefined are acceptable in terms of whether or not a crash happens.

So, it becomes a matter of intent.  Should we think of `text` as a glorified `toString` or do we intend to define the purpose of `text` in a more curated way?  @hlibco @jorgebucaran

Reply:
@icylace That's a good point. If `text` abstracts a native method that performs a `toString` under the hood, I think this feature should be exposed upstream.

Here's an experiment using `innerText` to observe how it converts non-strings to strings.
https://codepen.io/hlibco/pen/NWxeQmo

As a reference: https://dom.spec.whatwg.org/#text

Reply:
@hlibco I think no changes to `text` are required for the time being.  Would you agree?